starfish/intensity_table.py

```diff
@@ -276,3 +278,138 @@ def synthetic_intensities(
+        return mask
+
+    def mask_small_features(self, size_threshold):
+        """return the indices of features whose radii are smaller than size_threshold"""
+        mask = np.where(self.coords.features[self.SpotAttributes.RADIUS.value] < size_threshold)[0]
+        return mask
+
+    def _intensities_from_regions(self, props, reduce_op='max') -> "IntensityTable":
+        """turn regions back into intensities by reducing over the labeled area"""
+        raise NotImplementedError
+
+    def combine_adjacent_features(
+            self, area_threshold, assume_contiguous=True
+    ) -> Tuple["IntensityTable", List, np.ndarray, np.ndarray, np.ndarray]:
+        """
+
+        Parameters
+        ----------
+        area_threshold
+        assume_contiguous
+
+        Notes
+        -----
+
+        --
+        # This is how a non-contiguous ImageStack creation could work:
+        # construct an empty (z, y, x) image
+        decoded_image = np.zeros(self.attrs['image_shape'], dtype=int)
+
+        # fill it with features
+        coordinates = self.indexes['features'].to_frame()[['z', 'y', 'x']]
+        genes = self.gene_name.values
+
+        for i in np.arange(self.sizes[self.Constants.FEATURES.value]):
+            z, y, x = coordinates.iloc[i]
+            decoded_image[z, y, x] = gene_to_int[genes[i]]
+
+        --
+        # ideally we would be able to define a function that would extract the original
+        # area of the now-merged spot and calculate characteristics of it (max intensity, etc)
+        # across the code. It's signature could look a bit like this:
+        def props_to_intensities(props, label_image, image_stack, reduce_function
+                                 ) -> "IntensityTable":
+            tiles: List[np.ndarray] = []
+            for prop in props:
+                # extract the relevant area of the table and apply the reduce function
+                raise NotImplementedError
+
+        Returns
+        -------
+        IntensityTable :
+            spot intensities
+        List[
+
+        """
+        if not assume_contiguous:
+            raise NotImplementedError
+
+        # None needs to map to zero, non-none needs to map to something else.
+        int_to_gene = dict(
+            zip(range(1, np.iinfo(np.int).max),
+                set(self.gene_name.values) - {'None'}))
+        int_to_gene[0] = 'None'
+        gene_to_int = {v: k for (k, v) in int_to_gene.items()}
+
+        # map genes to ints
+        gene_list = [gene_to_int[g] for g in self.coords[self.Constants.GENE.value].values]
+        gene_array = np.array(gene_list)
+        decoded_image = gene_array.reshape(self.attrs['image_shape'])  # reverse linearization
+
+        # label the image and extract max intensity across each feature
+        label_image = label(decoded_image, connectivity=2)
+        intensities = self.max(dim=[Indices.CH, Indices.HYB])  # todo can use mean here too
```

Comment:
I think this supposed to be the magnitude of the pixel traces. In other words, for each pixel, it's the magnitude of the pixel trace.